File: app/src/main/java/app/we/go/emojidraw/arch/di/ServiceModule.java
package app.we.go.emojidraw.arch.di;

import javax.inject.Singleton;

import app.we.go.emojidraw.api.ApiConstants;
import app.we.go.emojidraw.api.EmojiDetectionService;
import dagger.Module;
import dagger.Provides;
import okhttp3.OkHttpClient;
import okhttp3.logging.HttpLoggingInterceptor;
import retrofit2.Retrofit;
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory;
import retrofit2.converter.gson.GsonConverterFactory;

/**
 * Provides dependencies related to the remote image recognition service
 */
@Singleton
@Module
class ServiceModule {

    public ServiceModule() {
    }

    @Provides
    @Singleton
    Retrofit provideRetrofit() {
        final HttpLoggingInterceptor loggingInterceptor = new HttpLoggingInterceptor();
        loggingInterceptor.setLevel(HttpLoggingInterceptor.Level.BODY);
        OkHttpClient client = new OkHttpClient.Builder()
                .addInterceptor(loggingInterceptor)
                .build();

        //noinspection UnnecessaryLocalVariable
        final Retrofit retrofit = new Retrofit.Builder()
                .client(client)
                .baseUrl(ApiConstants.Companion.getBASE_URL())
                .addConverterFactory(GsonConverterFactory.create())
                .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
                .build();

        return retrofit;
    }

    @Provides
    @Singleton
    EmojiDetectionService provideEmojiDetectionService(Retrofit retrofit) {
        return retrofit.create(EmojiDetectionService.class);
    }



}

File: app/src/main/java/app/we/go/emojidraw/arch/di/ServiceModule.kt
package app.we.go.emojidraw.arch.di

import app.we.go.emojidraw.api.BASE_URL
import app.we.go.emojidraw.api.EmojiDetectionService
import dagger.Module
import dagger.Provides
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Retrofit
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory
import retrofit2.converter.gson.GsonConverterFactory
import javax.inject.Singleton

@Module
class ServiceModule {

    @Provides
    @Singleton
    fun provideRetrofit(): Retrofit {
        val loggingInterceptor = HttpLoggingInterceptor()
        loggingInterceptor.level = HttpLoggingInterceptor.Level.BODY
        val client = OkHttpClient.Builder()
                .addInterceptor(loggingInterceptor)
                .build()

        return Retrofit.Builder()
                .client(client)
                .baseUrl(BASE_URL)
                .addConverterFactory(GsonConverterFactory.create())
                .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
                .build()
    }


    @Provides
    @Singleton
    fun provideEmojiDetectionService(retrofit: Retrofit): EmojiDetectionService {
        return retrofit.create(EmojiDetectionService::class.java)
    }


}

